allennlp/training/wandb_writer.py

```diff
@@ -0,0 +1,127 @@
+
+from overrides import overrides
+import torch
+
+from allennlp.common import Params
+from allennlp.models.model import Model
+from allennlp.training.log_writer import LogWriter
+from allennlp.training.optimizers import Optimizer
+
+
+@LogWriter.register("wandb")
+class WandBWriter(LogWriter):
+    """
+    Logs training runs to Weights & Biases.
+
+    !!! Note
+        This requires the environment variable 'WANDB_API_KEY' to be set.
+
+    In addition to the parameters that `LogWriter` takes, there are several other
+    parameters specific to `WandBWriter` listed below.
+
+    # Parameters
+
+    project : `Optional[str]`, optional (default = `None`)
+        The name of the W&B project to save the training run to.
+    tags : `Optional[List[str]]`, optional (default = `None`)
+        Tags to assign to the training run in W&B.
+    watch_model : `bool`, optional (default = `True`)
+        Whether or not W&B should watch the `Model`.
+    files_to_save : `Tuple[str, ...]`, optional (default = `("config.json", "out.log")`)
+        Extra files in the serialization directory to save to the W&B training run.
+    """
+
+    def __init__(
+        self,
+        serialization_dir: str,
+        model: Model,
+        optimizer: Optimizer,
+        summary_interval: int = 100,
+        distribution_interval: Optional[int] = None,
+        batch_size_interval: Optional[int] = None,
+        should_log_parameter_statistics: bool = True,
+        should_log_learning_rate: bool = False,
+        project: Optional[str] = None,
+        tags: Optional[List[str]] = None,
+        watch_model: bool = True,
+        files_to_save: Tuple[str, ...] = ("config.json", "out.log"),
+    ) -> None:
+        if "WANDB_API_KEY" not in os.environ:
+            raise ValueError("Missing environment variable 'WANDB_API_KEY'")
+
+        super().__init__(
+            serialization_dir,
+            model,
+            optimizer,
+            summary_interval=summary_interval,
+            distribution_interval=distribution_interval,
+            batch_size_interval=batch_size_interval,
+            should_log_parameter_statistics=should_log_parameter_statistics,
+            should_log_learning_rate=should_log_learning_rate,
+        )
+
+        self._files_to_save = files_to_save
+
+        import wandb
+
+        self.wandb = wandb
+        config_path = os.path.join(self._serialization_dir, "config.json")
+        self.wandb.init(
+            dir=os.path.abspath(serialization_dir),
+            project=project,
+            config=Params.from_file(config_path).as_dict(),
+            tags=tags,
+        )
+
+        for fpath in self._files_to_save:
+            # NOTE: this just tells W&B to watch for these files in the run directory
+            # and upload them at the end. They don't have to exist at this moment,
+            # and they won't. In `.close()`, we symlink these files from the serialization_dir
+            # to the run directory.
```

Comment:
Does that mean we see no metrics until the run is finished?

Reply:
No the metrics we log are uploaded on the fly

Reply:
Ah, only the artifacts get uploaded at the end? That could take a long time, right?

If something goes wrong while uploading, can you run with `--recover` and get the upload to finish?

Reply:
Hmm actually I guess they are uploaded as soon as the symlink is created, and then whenever the files linked to update afterwards. I'll make a change so that the symlinks are created on initialization, so the files will also be uploaded on the fly during training.

I'm still not sure what would happen if the files failed to upload before training finished.

Reply:
Can you know which files we will have when training starts?